### ISSUE TITLE
[Feature Request] Utilities for batch qMC (Sobol) sampling

### ISSUE THREAD

**Issue by saitcakmak** [pytorch/botorch#507]
# 🚀 Feature Request

The current qMC sampler `draw_sobol_samples` only allows for generating samples of shape `n x q x d`. It would be nice to have a method that generates low discrepancy samples of shape `n x *batch_shape x q x d` for an arbitrary batch shape. The desired output is equivalent to running a for loop over batches to generate `n x q x d` Sobol samples for each batch. 

## Motivation

The use case for this is generating the `raw_samples` / `initial_conditions` for optimization of batched GP models / acquisition functions. One can use `torch.rand` to generate `n x batch_shape x q x d` samples, however, having low discrepancy samples is more desirable for obvious reasons.

## Pitch

**Describe the solution you'd like**

A method `draw_batch_sobol_samples(bounds, n, q, batch_shape, seed)` that returns a `n x *batch_shape x q x d` tensor of samples where each batch is an independent `n x q x d` tensor of Sobol samples. The method should avoid initializing multiple `SobolEngine`, as this leads to huge overhead; as well as reshaping a big batch of Sobol samples, as this leads to #504.

**Describe alternatives you've considered**

- A seemingly obvious solution is to generate `(n * batch_size) x q x d` samples using `draw_sobol_samples`, and reshape it. This leads to the buggy samples described in #504. 
- Another obvious, though, inefficient solution is to generate `n x q x d` samples for each batch using `draw_sobol_samples` in a for loop. This leads to huge overhead due to repeated initialization of `SobolEngine`.
- Another alternative is to initialize the `SobolEngine` once, use this within a loop to generate `n x q x d` samples for each batch. To avoid issues similar to #504, after each call to `SobolEngine.draw(n)`, call `SobolEngine.fast_forward(torch.randint(some_constant), (1,))`. 
    This avoids overhead due to repeated initialization of `SobolEngine` and possibly #504. Using a for loop may get expensive as `batch_size` increases, and it may possibly lead to some unforeseen issues similar to #504. 
    It may be possible to achieve a better performance by directly modifying the `SobolEngine` source code.
    Repeated calls to `SobolEngine.draw` can also be avoided by first generating the full sequence of `randint` that are passed to `fast_forward`; calling `draw(n * batch_size + sum(randint))` to generate all samples at once; then deleting the samples corresponding to `fast_forward` before reshaping the samples to  `n x *batch_shape x q x d`.

**Are you willing to open a pull request?** (See [CONTRIBUTING](../CONTRIBUTING.md))
Yes.

**Comment by Balandat** [comment]
Having this functionality available makes sense to me. From an API perspective, we could just make `batch_shape` an optional arg to `draw_sobol_samples`, then we wouldn't need a new function for this.

> Repeated calls to SobolEngine.draw can also be avoided by first generating the full sequence of randint that are passed to fast_forward; calling draw(n * batch_size + sum(randint)) to generate all samples at once; then deleting the samples corresponding to fast_forward before reshaping the samples to n x *batch_shape x q x d.

Not quite sure I fully understand what you're suggesting here, mind elaborating a bit? 

How much is the overhead of initializing the SobolEngine vs. drawing the samples? Do you have some timing results profiling this? I'd like to understand how big of an issue that is (and what kinds of batch sizes you usually work with).

Python looping is kind of slow, this could be done much faster with SobolEngine in C++. But doing that is something we'd have to take on upstream, and would require quite a bit of work.

**Comment by saitcakmak** [comment]
> From an API perspective, we could just make batch_shape an optional arg to draw_sobol_samples, then we wouldn't need a new function for this.
Yes, this makes more sense.

> Not quite sure I fully understand what you're suggesting here, mind elaborating a bit?
I'll add a quick implementation in a bit. It should be easier that way.

> How much is the overhead of initializing the SobolEngine vs. drawing the samples? Do you have some timing results profiling this? I'd like to understand how big of an issue that is (and what kinds of batch sizes you usually work with).
Below are the times (in seconds) for 1000 replications each. I'm attaching the script just in case.
```
initialization with dim=10: 0.524356
drawing n=1000 samples of dim=10s: 0.026776
initialization with dim=100: 4.402194
drawing n=1000 samples of dim=100: 0.132297
initialization with dim=1000: 44.751162
drawing n=1000 samples of dim=1000: 1.218408
```
My typical batch size would be something like `[64, 10]` or `[64, 100]`. I'd be generating samples of shape e.g. `100 x 64 x 100 x 1 x 4`.

**Comment by saitcakmak** [comment]
Here is the proposed method:
```
import torch
from torch import Tensor
from typing import Optional, Union
from torch.quasirandom import SobolEngine


def draw_sobol_samples(
        bounds: Tensor,
        n: int,
        q: int,
        batch_shape: Optional[Union[torch.Size, tuple, list]] = None,
        seed: Optional[int] = None
) -> Tensor:
    r"""Draw qMC samples from the box defined by bounds.

    Args:
        bounds: A `2 x d` dimensional tensor specifying box constraints on a
            `d`-dimensional space, where bounds[0, :] and bounds[1, :] correspond
            to lower and upper bounds, respectively.
        n: The number of (q-batch) samples.
        q: The size of each q-batch.
        batch_shape: The batch shape of the samples. If given, returns samples
            of shape `n x *batch_shape x q x d`.
        seed: The seed used for initializing Owen scrambling. If None (default),
            use a random seed.

    Returns:
        A `n [x *batch_size] x  q x d`-dim tensor of qMC samples from the
        box defined by bounds.

    Example:
        >>> bounds = torch.stack([torch.zeros(3), torch.ones(3)])
        >>> samples = draw_sobol_samples(bounds, 10, 2)
    """
    batch_shape = torch.Size([]) if batch_shape is None else batch_shape
    batch_size = int(torch.prod(torch.tensor(batch_shape)))
    d = bounds.shape[-1]
    lower = bounds[0]
    rng = bounds[1] - bounds[0]
    sobol_engine = SobolEngine(q * d, scramble=True, seed=seed)
    if batch_size == 1:
        # the old non-batch implementation
        samples_raw = sobol_engine.draw(n, dtype=lower.dtype)
    else:
        # the experimental batched implementation.
        # we will draw a large (n) number of samples, and delete a random number of
        # samples between each batch to avoid issues similar to #504.
        num_delete = torch.randint(min(100, n), (batch_size,))
        num_delete[0] = 0  # no need to delete before the first batch
        # number of samples to draw including the ones to be deleted
        num_draw = n * batch_size + torch.sum(num_delete)
        samples_raw = sobol_engine.draw(num_draw, dtype=lower.dtype)
        # we will now create the indices of samples to keep.
        final_idx = torch.arange(n * batch_size).view(batch_size, n)
        # sum up num_delete to get the skip amount before each batch
        num_skip = torch.matmul(
            torch.tril(torch.ones([batch_size, batch_size], dtype=torch.long)),
            num_delete.view(-1, 1)
        )
        # indices of samples to keep
        final_idx = final_idx + num_skip.repeat(1, n)
        # delete the samples in between batches.
        # The resulting tensor only includes n * batch_size samples
        samples_raw = samples_raw[final_idx.view(-1)]
    samples_raw = samples_raw.view(n, *batch_shape, q, d).to(device=lower.device)
    return lower + rng * samples_raw
```
Let's say I need to generate `5` batches of `10` samples each, and `num_delete = [0, 2, 3, 2, 1]`. I generate `56` samples, assign first `10` to first `batch`, skip `2`, assign `[12: 22]` to second batch, skip `3` assign `[25:35]` etc.

**Comment by saitcakmak** [comment]
Looks like this introduces a non-negligible overhead, particularly for large batches and small `n`.
For generating `1000 x 64 x 100 x 1 x 4'
```
old buggy draw: 11.758210
new batch draw: 48.068122
```


For `100 x 64 x 100 x 1 x 4'
```
old buggy draw: 1.014645
new batch draw: 33.110637
```

For `100 x 64 x 100 x 1 x 4'
```
old buggy draw: 0.940627
new batch draw: 1.767405
```

Timeit code:
```
from timeit import timeit

number = 100

setup = """
import torch
from debug_etc.batch_sobol import draw_sobol_samples
dim = 4
bounds = torch.tensor([[0.], [1.]]).repeat(1, dim)
n = 1000
q = 1
batch_shape = [64, 100]
batch_size = int(torch.prod(torch.tensor(batch_shape)))
"""
test = """
samples = draw_sobol_samples(bounds, n*batch_size, q)
samples = samples.reshape(n, *batch_shape, q, dim)
"""
time = timeit(setup=setup, stmt=test, number=number)
print("old buggy draw: %f" % time)

setup = """
import torch
from debug_etc.batch_sobol import draw_sobol_samples
dim = 4
bounds = torch.tensor([[0.], [1.]]).repeat(1, dim)
n = 1000
q = 1
batch_shape = [64, 100]
batch_size = int(torch.prod(torch.tensor(batch_shape)))
"""
test = "draw_sobol_samples(bounds, n, q, batch_shape)"
time = timeit(setup=setup, stmt=test, number=number)
print("new batch draw: %f" % time)
```

**Comment by saitcakmak** [comment]
I looked into the overhead using `cProfile`. It looks like the operation for doing the triangular sum, i.e. getting a tensor where each element is `new_tensor[i] = torch.sum(old_tensor[:i+1])`, accounts for the bulk of it. In the above code, this is achieved using
```
        num_skip = torch.matmul(
            torch.tril(torch.ones([batch_size, batch_size], dtype=torch.long)),
            num_delete.view(-1, 1)
        )
```
which avoids a for loop. The breakdown of runtime for generating `1000 x 64 x 100 x 1 x 6` samples (100 replications) is as follows:
# total 49.600
# _sobol_engine_draw 10.515
# torch.matmul 9.092
# torch.tril 16.253
# torch.ones 2.644
# torch.repeat 0.455
If we can speed up that step, then the overhead should be negligible

**Comment by saitcakmak** [comment]
I tried a few alternatives. Replacing 
```
        num_skip = torch.matmul(
            torch.tril(torch.ones([batch_size, batch_size], dtype=torch.long)),
            num_delete.view(-1, 1)
        )
```
with 
```
        num_skip = torch.tensor(
            [torch.sum(num_delete[:i+1]) for i in range(num_delete.shape[0])]
        )
```
reduces the runtime significantly. 
```
# total 25.643
# _sobol_engine_draw 10.563
# torch.sum 2.878
# torch.tensor 0.469
# torch.repeat 0.416
```

The run times from https://github.com/pytorch/botorch/issues/507#issuecomment-670032117 are updated as follows (all 100 replications):
For generating `1000 x 64 x 100 x 1 x 4'
```
old buggy draw: 11.758210
new batch draw : 48.068122
new batch draw w/ improvements: 21.398802
```

For `100 x 64 x 100 x 1 x 4'
```
old buggy draw: 1.014645
new batch draw: 33.110637
new batch draw w/ improvements: 6.684207
```

For `1000 x 64 x 10 x 1 x 4'
```
old buggy draw: 0.896566
new batch draw: 1.941403
new batch draw w/ improvements: 2.161677

```

For `100 x 64 x 100 x 100 x 4'
```
old buggy draw: 95.537857
new batch draw: 168.840732
new batch draw w/ improvements: 142.451254
```

For `100 x 4 x 10 x 10 x 4'
```
old buggy draw: 0.258014
new batch draw: 0.314790
new batch draw w/ improvements: 0.340382
```

**Comment by saitcakmak** [comment]
Actually, the samples generated this way are not necessarily low discrepancy. Here is the script I used to look into this 
[calc_discrepancy.txt](https://github.com/pytorch/botorch/files/5037941/calc_discrepancy.txt). Without any batches, if we simply look at the proportion of samples in `[0, 0.5]` and `[0.5, 1]`, these are equal. With batches, I saw values going up to `0.8` to `0.20` (which would be reported as a discrepancy of `0.6` in the script).

**Comment by saitcakmak** [comment]
@Balandat , ignore everything above. The problem was with the way I kept reshaping the samples. We do not need to throw away samples, nor call `draw` multiple times. Here is the proper implementation, with actual low-discrepancy samples, and almost no overhead.
```
def draw_sobol_samples(
        bounds: Tensor,
        n: int,
        q: int,
        batch_shape: Optional[Union[torch.Size, tuple, list]] = None,
        seed: Optional[int] = None
) -> Tensor:
    r"""Draw qMC samples from the box defined by bounds.

    Args:
        bounds: A `2 x d` dimensional tensor specifying box constraints on a
            `d`-dimensional space, where bounds[0, :] and bounds[1, :] correspond
            to lower and upper bounds, respectively.
        n: The number of (q-batch) samples.
        q: The size of each q-batch.
        batch_shape: The batch shape of the samples. If given, returns samples
            of shape `n x *batch_shape x q x d`.
        seed: The seed used for initializing Owen scrambling. If None (default),
            use a random seed.

    Returns:
        A `n [x *batch_size] x  q x d`-dim tensor of qMC samples from the
        box defined by bounds.

    Example:
        >>> bounds = torch.stack([torch.zeros(3), torch.ones(3)])
        >>> samples = draw_sobol_samples(bounds, 10, 2)
    """
    batch_shape = torch.Size([]) if batch_shape is None else batch_shape
    batch_size = int(torch.prod(torch.tensor(batch_shape)))
    d = bounds.shape[-1]
    lower = bounds[0]
    rng = bounds[1] - bounds[0]
    sobol_engine = SobolEngine(q * d, scramble=True, seed=seed)
    # the experimental batched implementation without throwing away any samples
    # we're just doing proper reshaping here
    # the idea is we just need to keep the consecutive samples together within a batch,
    # so that first n goes to first batch, second n to second batch etc.
    # Simply doing reshape(n, *batch_shape, q, d) will not do this. We need to reshape to
    # `*batch_shape x n x q x d` then permute.
    samples_raw = sobol_engine.draw(n * batch_size, dtype=lower.dtype)
    samples_raw = samples_raw.view(*batch_shape, n, q, d).to(device=lower.device)
    if batch_shape != torch.Size([]):
        samples_raw = samples_raw.permute(
            -3, *range(len(batch_shape)), -2, -1
        )
    return lower + rng * samples_raw
```

**Comment by saitcakmak** [comment]
I'm gonna clean it up and create a PR

**Comment by saitcakmak** [comment]
Closed by #510
